### PR TITLE
Add WaterDrop::Testing::SpecKafkaClient

### PR DESCRIPTION
**Pull request by id-ilych**
Add a helper client to use in tests to monitor produced messages.
It is based on [the similar class](https://github.com/karafka/karafka-testing/blob/master/lib/karafka/testing/spec_producer_client.rb) from `karafka-testing` but is simplified and is aimed for producer-only apps.

### CODE REVIEW

**Comment by mensfeld** [comment]
retargeting so I can apply some small remarks and maybe rename

**Comment by id-ilych** [comment]
Hey @mensfeld :wave:

I fixed offenses reported by coditsu. Is there any way to run the checks locally?

**Comment by mensfeld** [comment]
@nijikon can you help @id-ilych with Coditsu?

**Comment by id-ilych** [comment]
Hey @mensfeld 👋 

Can you please restart Coditsu? Tomasz seems to be unavailable.

**Comment by nijikon** [comment]
@id-ilych sorry, I missed this.

**Comment by id-ilych** [comment]
Hey @nijikon :wave:

Please re-run it.

**Comment by id-ilych** [comment]
Thank you @nijikon 🙇 

@mensfeld FYI CI is green, PR is ready to be merged :)

File: spec/lib/waterdrop/testing/spec_kafka_client_spec.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  before do
+    allow(producer).to receive(:client).and_return(client)
+
+    producer.produce_sync(payload: 'one', topic: 'foo')
+    producer.produce_sync(payload: 'one', topic: 'bar')
+    producer.produce_sync(payload: 'two', topic: 'foo')
+  end
+
+  describe '#messages' do
```
Comment (nijikon) [inline spec/lib/waterdrop/testing/spec_kafka_client_spec.rb:42]:
```ruby
  describe '#messages' do
    subject { client.messages }

    it { is_expected.to match(all_messages) }
  end
```

File: spec/lib/waterdrop/testing/spec_kafka_client_spec.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    it 'yields all produced messages' do
+      expect(client.messages).to match(all_messages)
+    end
+  end
+
+  describe '#messages_for' do
+    context 'with topic that has messages produced to' do
+      it 'yields corresponding messages' do
```
Comment (nijikon) [inline spec/lib/waterdrop/testing/spec_kafka_client_spec.rb:50]:
```ruby
it { expect(client.messages_for('foo')).to match(foo_messages) }
it { expect(client.messages_for('bar')).to match(bar_messages) }
```

File: spec/lib/waterdrop/testing/spec_kafka_client_spec.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        expect(client.messages_for('foo')).to match(foo_messages)
+        expect(client.messages_for('bar')).to match(bar_messages)
+      end
+    end
+
+    context 'when topic that was not produced to' do
+      it 'yields no messages' do
```
Comment (nijikon) [inline spec/lib/waterdrop/testing/spec_kafka_client_spec.rb:57]:
```ruby
it { expect(client.messages_for('buzz')).to be_empty }
```

File: spec/lib/waterdrop/testing/spec_kafka_client_spec.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        expect(client.messages_for('buzz')).to be_empty
+      end
+    end
+  end
+
+  describe '#reset' do
+    it 'clears all buffers' do
```
Comment (nijikon) [inline spec/lib/waterdrop/testing/spec_kafka_client_spec.rb:64]:
```ruby
before { client.reset }

it { expect(client.messages).to be_empty }
it { expect(client.messages_for('foo')).to be_empty }
```

File: spec/lib/waterdrop/testing_spec.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+
+# Nothing really here. Just a stub for Coditsu
```
Comment (nijikon) [inline spec/lib/waterdrop/testing_spec.rb:3]:
```ruby
RSpec.describe_current do
  pending
end
```

**Comment by nijikon** [comment]
@id-ilych added some comments. Let me know what you think.

File: spec/lib/waterdrop/testing_spec.rb (hunk 5 above)
Comment (id-ilych) [reply spec/lib/waterdrop/testing_spec.rb:3]: @nijikon I copied that file from `spec/lib/waterdrop/contracts_spec.rb` should that one be changed too?

**Comment by id-ilych** [comment]
@nijikon regarding the other suggestions, my **_subjective_** thoughts are:
1. Hand-written block titles produce more descriptive output (`--format=documentation`).
2. Introducing `subject` for `#messages` without introducing it for `#messages_for` next to it, feels inconsistent. And introducing it for `#messages_for` via sub-contexts and `let(:topic)` seems to be an overkill (too much indirection killing readability).

But the main thing here is the **_project-wise consistency_**, so assuming all your suggestions are aligned with the general guidelines for the project **_I applied_** those 👍

File: spec/lib/waterdrop/testing_spec.rb (hunk 5 above)
Comment (nijikon) [reply spec/lib/waterdrop/testing_spec.rb:3]: Ah, fair. If it is like that in this repository, then we are good.

**Comment by mensfeld** [comment]
Thanks guys. I will get back to this at the end of this week :pray:

**Comment by nijikon** [comment]
1. In some cases, it makes sense to introduce a specific text as to what you wrote. It makes it more descriptive. In this example, the default one brings more information.
```ruby
WaterDrop::Testing::SpecKafkaClient
  #messages_for
    with topic that has messages produced to it
      yields corresponding messages (FAILED - 1)
      is expected to match [{:payload=>"one1", :topic=>"bar"}] (FAILED - 2)
      is expected to match [{:payload=>"one1", :topic=>"foo"}, {:payload=>"two2", :topic=>"foo"}] (FAILED - 3)
```
2. I agree with you, subject is not needed.

Thank you for all the work you did here. Much appreciated.

**Comment by mensfeld** [comment]
Checking...

**Comment by mensfeld** [comment]
@id-ilych looks good. My proposition here (I'll do it) is to expand config with `client_type`: where default is `:rdkafka`, second is `:dummy` and third (yours) is `:buffer`. `#client` on producer is public anyhow so this would be easy to reference.

To keep backward compatibility `deliver` set to `false` would always set `:dummy` client in case `:regular` is configured. Otherwise would use the `:client_type`:

```ruby
WaterDrop::Producer.new do |config|
  config.deliver = false
  config.client_type = :buffer
end

WaterDrop::Producer.new do |config|
  config.deliver = false
end

WaterDrop::Producer.new do |config|
  config.client_type = :buffer
end
```

etc. Then we could just create a `WaterDrop::Clients` namespace and move all there (with regular just as a passthrough because it anyhow comes from the builder).

WDYT? That way we do not explicitly provide "testing" but  we expand and allow to use buffer for dev or anything else when connection to Kafka is not needed.

**Comment by id-ilych** [comment]
Sounds great! :+1:

**Review by mensfeld — APPROVED** [review]

**Pull request merged**